prov/util/src/util_mr_cache.c

```diff
@@ -72,29 +73,13 @@ static int util_mr_find_overlap(struct ofi_rbmap *map, void *key, void *data)
 	return 0;
 }
 
-static struct ofi_mr_entry *util_mr_entry_alloc(struct ofi_mr_cache *cache)
-{
-	struct ofi_mr_entry *entry;
-
-	pthread_mutex_lock(&cache->lock);
-	entry = ofi_buf_alloc(cache->entry_pool);
-	pthread_mutex_unlock(&cache->lock);
-	return entry;
-}
-
-static void util_mr_entry_free(struct ofi_mr_cache *cache,
-			       struct ofi_mr_entry *entry)
-{
-	pthread_mutex_lock(&cache->lock);
-	ofi_buf_free(entry);
-	pthread_mutex_unlock(&cache->lock);
-}
-
 /* We cannot hold the monitor lock when freeing an entry.  This call
  * will result in freeing memory, which can generate a uffd event
  * (e.g. UNMAP).  If we hold the monitor lock, the uffd thread will
  * hang trying to acquire it in order to read the event, and this thread
  * will itself be blocked until the uffd event is read.
+ *
+ * Must hold cache lock.
```

Comment:
We cannot hold a lock around the add/delete_region() callbacks.  This is what causes the hangs as that results in memory allocations/deallocations.